### ISSUE TITLE
Move training and evaluation related functions from module functions to member functions

### ISSUE THREAD

**Issue by chunyang-wen** [sql-machine-learning/elasticdl#1006]
Currently in order to train models, ElasticDL training worker requires following command line parameters:
+ `model_def`:
  + Name of the class which inherits from `tf.keras.Model` such as `mnist_module.MnistModel`.
  + Name of the function which returns a `tf.keras.Model` instance such as `mnist_module.create_model`
+ `optimizer`: A function that returns a tensorflow optimizer such as `tf.train.GradientDescentOptimizer`
+ `loss`: A function that returns a tensor which represents loss.
+ `eval_metrics_fn`:  A function that returns a dict. The key of the dict is the metric's name and the value is the metric's cooresponding tensor.
+ `dataset_fn`: A function that returns a `tf.data.Dataset`. It is the input of training.

Those parameters are names of function in the module where we define `model_def`.  ElasticDL uses those functions in different stages of training or evaluation. Those functions are defined as module functions. It has following limitations:
+ It is hard to share information between functions unless we use global variables.
+ It is hard to find out that one of those functions is not defined until ElasticDL calls it. And it is too late at that time because training or evaluation process may have been running for a long time. 

As a result, we prefer a unified class to encapsulate them.  Let's call it `TrainSpec`.

```python
class TrainSpec(object):
      def __init__(self, context=None):
          """
          Args:
              context: a dict of contextual information from ElasticDL,
                             such as worker_id, master ip.
          """
          self._context = context

     @abstractmethod
     def create_optimizer(self, lr=0.1):
         """returns a tensorflow optimizer such as `tf.train.GradientDescentOptimizer`"""
         pass

     @abstractmethod
     def create_loss(self, outputs, labels):
        """returns a tensor represents loss using outputs and labels"""
        pass

     @abstractmethod
     def create_metrics(self,
                         mode=Mode.TRAINING,
                         outputs=None,
                         labels=None,
                         predictions=None,):
          """returns a dict of metrics"""
          pass

     @abstractmethod
     def create_model(self):
          """returns model created
          Users can use functional API or subclass to create customized Keras model.
          """

class MnistTrainSpec(TrainSpec):
    def create_optimizer(self, lr=0.1):
        return tf.trian.GradientDescentOptimizer(0.1)
    # other functions are left undefined purposely.
```
Users define `MnistTrainSpec` which inherits from `TrainSpec`. We will have at least two benefits:
+ We can share Information between functions as they are all member functions of the same class.
+ We can use abstract methods to help users to find out functions we require which are not implemented. ElasticDL will report `NotImplementedError` at the first time we try to instantiate the above `MnistTrainSpec`.

### Another design of `TrainSpec`

The above `TrainSpec` assumes that different `create_*` functions are isolated. If we need to share information between those functions, we have to put them in the constructor. Another drawback is that multiple calls to the same `create_*` functions may return different instances. For example `create_model` will return two instances of `tf.keras.Model` if we call it twice.

Why we run into this dilemma ? Because we have done too more work for users. Users themselves have better understanding of their model/loss/optimizer. They know when to create an optimizer, what information to share between different functions such as loss and model. 

What ElasticDL cares is only the results:
+ the model instance
+ the optimizer instance
+ the loss function
+ the dataset function
+ the metric function

So we redefine `TrainSpec` as an abstract base class with only abstract properties. If we want to train a mnist model, we can provide a class `MnistSpec`. It implements all the abstract properties. Users can use [functional API](https://www.tensorflow.org/guide/keras#functional_api) or [subclass](https://www.tensorflow.org/guide/keras#model_subclassing) to implement all those properties and return them.

```python
import abc
class TrainSpec(abc.ABC):
     def __init__(self, context=None):
     """
     Args:
         context: A dict of contextual information from ElasticDL, such as worker_id, master ip
     """
     self._context = context or {}

     @property
     @abstractmethod
    def optimizer(self):
        """returns an instance of tensorflow optimizer, such as `tf.train.GradientDescentOptimizer`"""

     @property
     @abstractmethod
    def loss_fn(self):
        """returns a loss function with parameters of `outputs` and `label`
       def loss_fn(outputs, labels):
              return tf.losses.mean_squared_error(labels, outputs)
       You can return  a function like`loss_fn`
        """
       pass

     @property
     @abstractmethod
      def model(self):
        """returns an instance of `tf.keras.Model` or instance of its subclass"""
       pass
  
     @property
     @abstractmethod
     def dataset_fn(self):
     """returns a function that generates a tf.data.Dataset instance
     def dataset_fn(files):
         # blabla
         return dataset
     You can return a dataset function like `dataset_fn`.
     """

     @property
     @abstractmethod
     def metrics_fn(self):
        """returns an metric function with parameters
         def metrics_fn(self,
                         mode=Mode.TRAINING,
                         outputs=None,
                         labels=None,
                         predictions=None,):
             if mode == Mode.EVALUATION:
                 return {"mse": tf.metrics.accuracy(labels, predictions)}
             else:
                 return {}
         You can return a function like metrics_fn
        """
        pass


class MnistModel(tf.keras.Model);
     def call(self, inputs, training=False):
         pass

def _loss_fn(outputs, labels):
    retrurn tf.losses.mean_square_error(labels, outputs)

class MnistSpec(TrainSpec)：
      def __init__(self, context=None):
          super(MnistSpec, self).__init__(context=context)
          self._optimizer = tf.train.GradientOptimizer()
          self._model = MnistModel()
      
      @property
      def optimizer(self):
           return self._optimizer

      @property
      def model(self):
           return self._model

      @property
      def loss_fn(self)
          return _loss_fn
      # other properties are left undefined purposely.
```
Users can create model/optimizer/loss_fn/dataset_fn anywhere. The above code defines variable used by property directly in `MnistSpec`'s constructor.

This new design will give us the same instance even if we access any property multiple times.
